### PR TITLE
Issue 1270: add checks for meta includes

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary

The main change of this PR is the addition of checks for meta includes in test-math-dependencies. It searches for `'<stan/math/.*/meta/.*hpp' `inside the `stan/math`, excluding all meta subfolders (`stan/math/*/*/meta/*`) and the `*/meta.hpp` files.
In order to support excluding files in the check I extended the math_warnings function.

I then fixed all the warnings that came out:

```
stan/math/prim/mat/prob/categorical_logit_glm_lpmf.hpp(4): File includes */meta/*.hpp header file. Should include meta.hpp [meta]
stan/math/prim/mat/prob/categorical_logit_glm_lpmf.hpp(8): File includes */meta/*.hpp header file. Should include meta.hpp [meta]
stan/math/prim/mat/prob/categorical_logit_glm_lpmf.hpp(9): File includes */meta/*.hpp header file. Should include meta.hpp [meta]
stan/math/prim/mat/prob/categorical_logit_glm_lpmf.hpp(10): File includes */meta/*.hpp header file. Should include meta.hpp [meta]
stan/math/prim/mat/prob/categorical_logit_glm_lpmf.hpp(11): File includes */meta/*.hpp header file. Should include meta.hpp [meta]
stan/math/prim/mat/prob/categorical_logit_glm_lpmf.hpp(12): File includes */meta/*.hpp header file. Should include meta.hpp [meta]
stan/math/prim/mat/prob/categorical_logit_glm_lpmf.hpp(13): File includes */meta/*.hpp header file. Should include meta.hpp [meta]
stan/math/prim/mat/prob/categorical_logit_glm_lpmf.hpp(14): File includes */meta/*.hpp header file. Should include meta.hpp [meta]
stan/math/fwd/core/fvar.hpp(4): File includes */meta/*.hpp header file. Should include meta.hpp [meta]
stan/math/fwd/core/fvar.hpp(6): File includes */meta/*.hpp header file. Should include meta.hpp [meta]
make/test-math-dependencies:48: recipe for target 'test-math-dependencies' failed
make: *** [test-math-dependencies] Error 1
```

In order to fix the error in `fwd/core/fvar.hpp` I had to move around includes in `fwd/meta.hpp`. That was just a hunch and more of trial and error solution than me understanding why that works. So I would kindly ask @drezap to double check if what I did is OK. The `test/unit/math/fwd/core/` tests pass.

## Tests

Add a include of a specific file from `meta/*.hpp` to a file not in a `/meta/` subfolder and run `make test-math-dependencies`. It should fail.

Example: 

Replace `#include <stan/math/prim/meta.hpp>` in `stan/math/fwd/core/fvar.hpp` with `#include <stan/math/prim/scal/meta/likely.hpp>` and `run make test-math-dependencies`

## Side Effects

/

## Checklist

- [x] Math issue #1270 

- [x] Copyright holder: Rok Češnovar, Univ. of Ljubljana

- [x] the basic tests are passing

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: make/test-math-dependencies [hunk 1]
```diff
@@ -24,7 +24,10 @@
 ##
 define math_warnings
 $(eval type:= $1)
-$(eval all_headers := $(shell find $2 -name '*.hpp' -o -name '*.cpp'))
+$(eval excl:=)
+$(eval excl_ids:= 1 2 3 4 5)
+$(foreach a,$(excl_ids),$(eval excl+=$(if $(word $(a),$5),! -path $(word $(a),$5),)))
+$(eval all_headers := $(shell find $2 -name '*.hpp' $(excl)-o -name '*.cpp'))
```
Comment (rok-cesnovar) [inline make/test-math-dependencies]: This changes supports up to 5 exclude options in the fifth argument. If the fifth argument is 'option1' 'option2' the find command is `shell find $2 -name '*.hpp' ! -path option1 ! -path option2 -o -name '*.cpp'`

File: stan/math/fwd/core/fvar.hpp [hunk 2]
```diff
@@ -1,9 +1,9 @@
 #ifndef STAN_MATH_FWD_CORE_FVAR_HPP
 #define STAN_MATH_FWD_CORE_FVAR_HPP
 
-#include <stan/math/prim/scal/meta/likely.hpp>
+#include <stan/math/prim/meta.hpp>
 #include <stan/math/prim/scal/fun/is_nan.hpp>
-#include <stan/math/fwd/scal/meta/ad_promotable.hpp>
+#include <stan/math/fwd/meta.hpp>
 #include <ostream>
```
Comment (drezap) [inline stan/math/fwd/core/fvar.hpp]: Thanks for doing this.

But the order looks good: first prim/meta, then any prim includes, then fwd includes.

File: stan/math/fwd/meta.hpp [hunk 3]
```diff
@@ -1,11 +1,10 @@
...
 #define STAN_MATH_FWD_META_HPP
 
+#include <stan/math/fwd/scal/meta/ad_promotable.hpp>
 #include <stan/math/fwd/scal/meta/is_fvar.hpp>
 #include <stan/math/fwd/scal/meta/partials_type.hpp>
-#include <stan/math/fwd/mat/meta/operands_and_partials.hpp>
-
-#include <stan/math/fwd/scal/meta/ad_promotable.hpp>
 #include <stan/math/fwd/scal/meta/operands_and_partials.hpp>
+#include <stan/math/fwd/mat/meta/operands_and_partials.hpp>
```
Comment (drezap) [inline stan/math/fwd/meta.hpp]: Yes, scal should be before mat.

File: stan/math/fwd/meta.hpp [hunk 4]
```diff
@@ -1,11 +1,10 @@
 #ifndef STAN_MATH_FWD_META_HPP
 #define STAN_MATH_FWD_META_HPP
 
+#include <stan/math/fwd/scal/meta/ad_promotable.hpp>
```
Comment (drezap) [inline stan/math/fwd/meta.hpp]: Yes, just like in the prim/meta.hpp, ad_promotable.hpp comes at the top. Good catch.

**Comment by drezap** [comment]
I'm not really familiar enough with this to do an official review, but include order changes look good.

File: stan/math/fwd/meta.hpp (hunk 4 above)
Comment (rok-cesnovar) [reply stan/math/fwd/meta.hpp]: Great, thanks!

**Comment by rok-cesnovar** [comment]
@drezap Thanks! I wanted your input on that, for the rest (the test-math-dependencies file) I am going to ask @syclik to review.

File: make/test-math-dependencies (hunk 1 above)
Comment (rok-cesnovar) [reply make/test-math-dependencies]: This has been resolved with a general foreach statement that now supports any number of exclude options as the fifth argument of the math-warnings function.

**Comment by rok-cesnovar** [comment]
@SteveBronder I generalized the check so that we can now add any number of exclude folders. 

The make help on test-math-dependencies now outputs the following: 

```
  - test-math-dependencies : walks through all the header files and indicates
      when the math dependencies are violated. Dependencies should follow:
      * rev -> prim
      * fwd -> prim
      * mix -> {rev, fwd, prim}
      * within {prim, rev, fwd, mix}: mat -> arr -> scal
      * only include {prim, rev, fwd, mix}/meta.hpp from the meta subfolders
```

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal I am seeing the following error on the "Windows Threading" test: 

```
Recording test results
No test report files were found. Configuration error?
No test report files were found. Configuration error?
```
Do you maybe have an idea on what could be the cause? I highly doubt its the PR.

**Comment by serban-nicusor-toptal** [comment]
At first sight I have no idea, I'm looking into this.

**Comment by SteveBronder** [comment]
@serban-nicusor-toptal anything we can help out with to get this running?

**Comment by serban-nicusor-toptal** [comment]
I've looked into it and the only difference between a successful run of this job and a failure is:

Good: https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/PR-1271/6/pipeline/146/

Bad: https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/PR-1271/11/pipeline/146

---

There is no output in `runTests.py -j25 test/unit -f thread --make-only` nor `runTests.py -j25 test/unit -f thread`.  
Hmh, I'm not sure if this is caused by the Jenkins itself.
I'm now running the job again with some minor modifications to see if it does the job.

**Comment by serban-nicusor-toptal** [comment]
Nah, it didn't do the trick.
Do you know what may have caused it to not output the logs ? On Jenkinsfile side there isn't anything changed that may affect this.

**Comment by SteveBronder** [comment]
Let me update this to master when I get home. Maybe this doesn't have the new windows threading stuff merged into it

**Comment by rok-cesnovar** [comment]
I merged develop in, hopefully this fixes it.

**Comment by wds15** [comment]
@serban-nicusor-toptal Here we have one more instance where Windows tests all fail.

**Comment by rok-cesnovar** [comment]
@SteveBronder your review was dissmised, please see the last commit. I had to fix a few more includes that were merged since this PR was reviewed.

**Pull request merged**